sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala

```diff
@@ -87,9 +87,12 @@ case class BroadcastExchangeExec(
             val beforeCollect = System.nanoTime()
             // Use executeCollect/executeCollectIterator to avoid conversion to Scala types
             val (numRows, input) = child.executeCollectIterator()
-            if (numRows >= 512000000) {
+            // Since the maximum number of keys that BytesToBytesMap supports is 1 << 29,
+            // and only 70% of the slots can be used before growing in HashedRelation,
+            // here the limitation should not be over 341 million.
+            if (numRows >= (1 << 29) / 1.5) {
               throw new SparkException(
-                s"Cannot broadcast the table with 512 million or more rows: $numRows rows")
+                s"Cannot broadcast the table with 341 million or more rows: $numRows rows")
```

Comment:
Maybe compute this limit in bytes and refer to it in the check and message, to be more specific

Reply:
Yea, +1 on the Sean's suggestion. btw, could we make this error message clearer? I personally think the current message looks unobvious for users.

Reply:
The limitation is `The maximum number of keys that BytesToBytesMap supports`. Its unit is not byte.

Reply:
Oops yes I just meant 'rows' or whatever. I just mean you can compute and reuse it in the error.

Reply:
@maropu Do you think we need to explain why uses 351m instead of 512m in the error message? I think it's just a limitation, whatever the value is. It just telling user a job fails due to broadcasting an unexpected large table.

Reply:
@srowen Ah, I get it.